src/main/java/gregtech/api/capability/impl/DistillationTowerLogicHandler.java

```diff
@@ -0,0 +1,218 @@
+package gregtech.api.capability.impl;
+
+import gregtech.api.capability.IMultipleTankHandler;
+import gregtech.api.metatileentity.multiblock.IMultiblockAbilityPart;
+import gregtech.api.metatileentity.multiblock.IMultiblockPart;
+import gregtech.api.metatileentity.multiblock.MultiblockAbility;
+import gregtech.api.pattern.BlockPattern;
+import gregtech.api.recipes.Recipe;
+import gregtech.api.util.GTLog;
+import gregtech.api.util.GTTransferUtils;
+import gregtech.common.metatileentities.multi.multiblockpart.MetaTileEntityMultiblockPart;
+
+import net.minecraft.util.math.BlockPos;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidTankInfo;
+import net.minecraftforge.fluids.IFluidTank;
+import net.minecraftforge.fluids.capability.FluidTankProperties;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+import net.minecraftforge.items.IItemHandlerModifiable;
+
+import com.cleanroommc.modularui.utils.FluidTankHandler;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DistillationTowerLogicHandler {
+
+    protected final AbstractRecipeLogic workable;
+    protected final IDistillationTower tower;
+
+    public int layerCount;
+    public List<IFluidHandler> orderedFluidOutputs;
+    public IMultipleTankHandler fluidTanks;
+
+    public DistillationTowerLogicHandler(IDistillationTower tower, AbstractRecipeLogic workable) {
+        this.tower = tower;
+        this.workable = workable;
+    }
+
+    protected boolean applyFluidToOutputs(List<FluidStack> fluids, boolean doFill) {
+        boolean valid = true;
+        for (int i = 0; i < fluids.size(); i++) {
+            IFluidHandler handler = orderedFluidOutputs.get(i);
+            int accepted = handler.fill(fluids.get(i), doFill);
+            if (accepted != fluids.get(i).amount) valid = false;
+            if (!doFill && !valid) break;
+        }
+        return valid;
+    }
+
+    public void outputRecipeOutputs() {
+        GTTransferUtils.addItemsToItemHandler(workable.getOutputInventory(), false, workable.itemOutputs);
+        this.applyFluidToOutputs(workable.fluidOutputs, true);
+    }
+
+    public boolean setupAndConsumeRecipeInputs(@NotNull Recipe recipe,
```

Comment:
Instead of this big copy paste, I think this should be reduced to only the changed part, and have implementers do their own copy paste replacing the one part they need with the contents of this method.

If `ARL#setupAndConsumeRecipeInputs` is modified later on, it's likely this case will be overlooked since it's not in the direct inheritance tree, so I think it would be better to have only direct overrides do the copy paste.